apiserver/facades/client/application/deploy.go

```diff
@@ -36,6 +36,7 @@ type DeployApplicationParams struct {
 	// instead of a machine spec.
 	Placement        []*instance.Placement
 	Storage          map[string]storage.Constraints
+	Devices          map[string]charm.Device
```

Comment:
We shouldn't introduce charm artefacts here. As with storage, we need to introduce a new struct. See juju/juju/storage/constraints.go